certexchange/polling/predictor.go

```diff
@@ -0,0 +1,108 @@
+package polling
+
+import (
+	"time"
+)
+
+const (
+	maxBackoffMultiplier = 10
+	minExploreDistance   = 100 * time.Millisecond
+)
+
+func newPredictor(targetAccuracy float64, minInterval, defaultInterval, maxInterval time.Duration) *predictor {
+	return &predictor{
+		minInterval:     minInterval,
+		maxInterval:     maxInterval,
+		interval:        defaultInterval,
+		exploreDistance: defaultInterval / 2,
+	}
+}
+
+// An interval predictor that tries to predict the time between instances. It can't predict the time
+// an instance will be available, but it'll keep adjusting the interval until we receive one
+// instance per interval.
+type predictor struct {
```

Comment:
I think this is plenty good as is. Way too early to optimise but writing down thoughts: In the future I suspect we can improve it further by piggybacking other peer's stats over exchanged certs request/response and feed it to a stochastic model to fine tune intervals.

Reply:
I think we have all the information we need locally and I'm weary of trusting what other peers tell us too much (too easy to bias/attack us). The main improvements I can see are:

1. Using the number of certificates. E.g., divide the current interval by the number of certificates received. I didn't do this yet because I don't want to adjust our prediction too quickly (especially if the interval is noisy).
2. Have some mechanism to discover how intervals should be _aligned_.

But... I don't think either actually matter.

1. I still want a pubsub based protocol which will allow us to keep up completely.
2. For GPBFT's purposes, we just need to be within 5 (the power table lookback) instances.

TBH, if we're _too accurate_, we'll constantly be making unnecessary network requests even when GPBFT is working perfectly and doesn't need them. That was actually the main issue with my previous attempt at this algorithm (it tried to find the exact "edge" where a new certificate would be released).

Reply:
Well, there are probably second-order facts we could learn to let us converge to the correct interval faster. E.g., given a sequence of events/choices, what choice should we make next that's most likely to find the optimal interval.

(we could probably straight-up use transformers for this...)